app/lib/frontend/backend.dart

```diff
@@ -860,6 +860,19 @@ Future<_ValidatedUpload> _parseAndValidateUpload(
 
   final files = await listTarball(filename);
 
+  // Check whether the files can be extracted on case-preserving file systems
+  // (e.g. on Windows). We can't allow two files with the same case-insensitive
+  // name.
+  final lowerCaseFiles = <String>{};
+  for (String file in files) {
+    final lower = file.toLowerCase();
+    if (lowerCaseFiles.contains(lower)) {
+      throw GenericProcessingException(
+          'Filename collision on case-preserving file systems: $file.');
+    }
+    lowerCaseFiles.add(lower);
```

Comment:
Nit, but `Set.add` does return a boolean :)
```dart
final lowerCaseFiles = <String>{};
for (String f in files) {
  if (!lowerCaseFiles.add(f.toLowerCase())) {
    throw GenericProcessingException(....);
  }
}
```
// https://api.dartlang.org/stable/2.2.0/dart-core/Set/add.html

Reply:
I find that negating add is less readable (I usually need to lookup what `add()` returns if item not already present). I find being explicit on the contains, but I don't mind changing it to add either...

Reply:
++ , it was just a nit... this is perfectly fine... :)